File: 프로그래머스/LV2/전력망을 둘로 나누기.py
from collections import deque


def div(s, e, visited, graph):
    q = deque([e])
    visited[s] = 1
    visited[e] = 1
    count = 1
    while q:
        cur = q.popleft()
        for next in graph[cur]:
            if not visited[next]:
                visited[next] = 1
                q.append(next)
                count += 1
    return count


def solution(n, wires):
    answer = 100
    graph = [[] for _ in range(n + 1)]
    for s, e in wires:
        graph[s].append(e)
        graph[e].append(s)

    for s, e in wires:
        visited = [0] * (n + 1)
        temp = div(s, e, visited, graph)
        ans = abs(temp - abs(n - temp))
        answer = min(ans, answer)
    return answer

File: 프로그래머스/javascript/LV2/전력망을 둘로 나누기.js
function solution(n, wires) {
    let answer = n;

    const graph = Array.from({ length: n + 1 }, () => []);
    for (const [s, e] of wires) {
        graph[s].push(e);
        graph[e].push(s);
    }

    function bfs(start, exclude) {
        const visited = Array(n + 1).fill(false);
        const queue = [start];
        visited[start] = true;
        visited[exclude] = true; // 제거한 간선의 반대편 노드 차단
        let count = 1;

        while (queue.length) {
            const cur = queue.shift();
            for (const next of graph[cur]) {
                if (!visited[next]) {
                    visited[next] = true;
                    queue.push(next);
                    count++;
                }
            }
        }
        return count;
    }

    for (const [s, e] of wires) {
        const count = bfs(s, e);
        answer = Math.min(answer, Math.abs(n - count * 2));
    }

    return answer;
}
